### PR TITLE
Profiles doc

### PR DESCRIPTION
**Pull request by mattias-p**
Partial fix for dotse/zonemaster#498.

I've made some assumptions here. Please check carefully.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
-Zonemaster Backend
-==================
+# Zonemaster Backend
+
 [![Build Status](https://travis-ci.org/dotse/zonemaster-backend.svg?branch=master)](https://travis-ci.org/dotse/zonemaster-backend)
 
+
 ### Purpose
 This repository is one of the components of the Zonemaster software. For an
 overview of the Zonemaster software, please see the
```
Comment (sandoche2k) [inline README.md]: Need to complete the sentence above.

File: docs/API.md [hunk 2]
```diff
@@ -140,19 +140,10 @@ The typical use case for this parameter would be a setup with several separate T
...
-
-One of the strings:
-
-* `"default_profile"`
-* `"test_profile_1"`
-* `"test_profile_2"`
-
-The `"test_profile_2"` *profile* is identical to `"default_profile"`.
+* Must consist entirely of letters A-Z and a-z, digits 0-9, hyphen '-' and underscore '_'.
+* Length must be between 1 and 32 characters inclusive.
```
Comment (matsduf) [inline docs/API.md]: "Must not start or end with '-' or '_'."

File: docs/Architecture.md [hunk 3]
```diff
@@ -71,9 +71,10 @@ soon as it can, and all of the real work will be done as that user.
...
 
->
-> TODO: Come up with a better name to distinguish it from *config profiles*.
->
+A profile is a configuration for Zonemaster Engine; see the [profiles overview] for context.
+Administrators can [configure available profiles], which is a mapping from [profile names] to [profiles files].
+The [RPC-API] contains several methods that accept profile name arguments.
+
 
 ### Config profile
```
Comment (matsduf) [inline docs/Architecture.md]: What is "Config profile"?

File: docs/Configuration.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+TBD
+
+
+## PROFILES section
+
+The PROFILES section declares the available [profiles].
+
+Keys are [profile names], and values are absolute file system paths to [profile JSON files].
+The profile named `default` is special and always available.
+When it isn't explicitly defined, it's implicitly mapped to the Zonemaster Engine default profile.
```
Comment (matsduf) [inline docs/Configuration.md]: My picture is the following:

1. The Engine default profile is always loaded.
2. If Backend defines profiles (symbolic name to a profile file), and a such a profile is selected, then that is used to override what is been defined by Engine (on top). If selected profile name or profile file is unavailable, backend will return an error.
3. Profile "default" can always be selected. If there is a profile file defined for "default", then that is loaded on on top of the Engine profile. If there is no such file defined, then that is just silently ignored.

**Comment by mattias-p** [comment]
@matsduf I'll get back to get_test_history once #320 is merged.

**Comment by mattias-p** [comment]
@mtoma Github says you left review comments, but now I can't find them. Could you please review this again and double check to see that Github actually displays your comments?

**Comment by mattias-p** [comment]
@matsduf and @mtoma Can you guys take another look at this PR to see if anything more needs to be done?

File: docs/Architecture.md [hunk 5]
```diff
@@ -67,24 +67,30 @@ soon as it can, and all of the real work will be done as that user.
...
-> TODO: Come up with a better name to distinguish it from *config profiles*.
->
+A profile is a configuration for Zonemaster Engine; see the [profiles
+overview] for context.
+Zonemaster Backend allows administrators to [configure the set of
+available profiles].
 
-### Config profile
+Each available profile has a [profile name].
+Each one is based on the [Zonemaster Engine default profile],
```
Comment (matsduf) [inline docs/Architecture.md]: Incorrect link or definition below "profile name" != "profile names"

File: docs/Architecture.md [hunk 6]
```diff
@@ -67,24 +67,30 @@ soon as it can, and all of the real work will be done as that user.
...
+Each one (with the possible exception of `default`) has a [profile file]
+with overrides to the Zonemaster Engine default profile.
```
Comment (matsduf) [inline docs/Architecture.md]: The default Engine default profile is always loaded. Any profile selected here will only override the parameters defined in the overriding profile file.

File: docs/Architecture.md [hunk 7]
```diff
@@ -67,24 +67,30 @@ soon as it can, and all of the real work will be done as that user.
...
 
 ### Engine
 
 The Zonemaster *Engine* is a library for performing *tests*. It's hosted in [its
 own repository](https://github.com/dotse/zonemaster-engine/).
+
+--------
+[Configure the set of available profiles]: Configuration.md#profiles-section
+[Profile files]: https://metacpan.org/pod/Zonemaster::Engine::Config#PROFILE-DATA
+[Profile names]: API.md#profile-name
```
Comment (matsduf) [inline docs/Architecture.md]: "profile names" or "profile name"?

**Comment by mattias-p** [comment]
@matsduf and @sandoche2k, I've addressed all of your comments.

File: docs/Configuration.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+The PROFILES section defines the available [profiles].
+
+Keys are [profile names], and values are absolute file system paths to
+[profile JSON files].
+Each profile JSON file contains a (possibly empty) set of overrides to
+the [Zonemaster Engine default profile].
+
+There is a `default` profile that is special.
+It is always available.
+If there isn't an overriding profile JSON file defined for the `default`
```
Comment (matsduf) [inline docs/Configuration.md]: A minor comment, but "is not" is better than "isn't" for this kind of text.

File: docs/Architecture.md [hunk 9]
```diff
@@ -67,24 +67,31 @@ soon as it can, and all of the real work will be done as that user.
...
+A profile is a configuration for Zonemaster Engine; see the [profiles
+overview] for context.
+Zonemaster Backend allows administrators to [configure the set of
+available profiles].
 
-### Config profile
+Each available profile has a [profile name].
+A profile named `default` is always available.
+Each available profile is based on the [Zonemaster Engine default profile].
+Each one (with the possible exception of `default`) has a [profile file]
```
Comment (matsduf) [inline docs/Architecture.md]: "[profile file]" --> "[profile files]" here or the reverse below.

**Comment by matsduf** [comment]
@sandoche2k and @mtoma, can you review this PR?

File: docs/Architecture.md (hunk 9 above)
Comment (mattias-p) [reply docs/Architecture.md]: Updated the link reference definition instead.

**Pull request merged**